src/Evolution/Systems/GrMhd/ValenciaDivClean/Flattener.cpp

```diff
@@ -0,0 +1,338 @@
+  bool already_computed_means = false;
+
+  const auto compute_means =
+      [&already_computed_means, &mean_tilde_d, &mean_tilde_tau, &mean_tilde_s,
+       &tilde_d, &tilde_tau, &tilde_s, &mesh, &det_logical_to_inertial_jacobian,
+       require_positive_mean_tilde_d =
+           require_positive_mean_tilde_d_]() {
+        if (already_computed_means) {
+          return;
+        }
+        already_computed_means = true;
+        // Compute the means w.r.t. the inertial coords
+        // (Note that several other parts of the limiter code take means w.r.t.
+        // the logical coords, and therefore might not be conservative on curved
+        // grids)
+        const double volume_of_cell =
+            definite_integral(get(det_logical_to_inertial_jacobian), mesh);
+        const auto inertial_coord_mean =
+            [&mesh, &det_logical_to_inertial_jacobian,
+             &volume_of_cell](const DataVector& u) {
+              // Note that the term `det_jac * u` below results in an
+              // allocation. If this function needs to be optimized, a buffer
+              // for the product could be allocated outside the lambda, and
+              // updated in the lambda.
+              return definite_integral(
+                         get(det_logical_to_inertial_jacobian) * u, mesh) /
+                     volume_of_cell;
+            };
+        mean_tilde_d = inertial_coord_mean(get(*tilde_d));
+        mean_tilde_tau = inertial_coord_mean(get(*tilde_tau));
+        for (size_t i = 0; i < 3; ++i) {
+          gsl::at(mean_tilde_s, i) = inertial_coord_mean(tilde_s->get(i));
+        }
+
+        if (require_positive_mean_tilde_d and mean_tilde_d < 0.) {
+          ERROR("Invalid TildeD input to flattener. Got "
+                << *tilde_d << " with mean value " << mean_tilde_d);
+        }
+      };
+
+  // If min(tilde_d) is negative, then flatten.
+  double min_d = std::numeric_limits<double>::max();
+  size_t location_of_min_d = std::numeric_limits<size_t>::max();
+  for (size_t i = 0; i < mesh.number_of_grid_points(); ++i) {
+    if (const double rest_mass_density_times_lorentz_factor =
+            get(*tilde_d)[i] / get(sqrt_det_spatial_metric)[i];
+        rest_mass_density_times_lorentz_factor < min_d) {
+      min_d = rest_mass_density_times_lorentz_factor;
+      location_of_min_d = i;
+    }
+  }
+  if (min_d < rest_mass_density_times_lorentz_factor_cutoff_) {
+    compute_means();
+
+    // Note: the current algorithm flattens all fields by the same factor,
+    // though in principle a different factor could be applied to each field.
+    //
+    // Experiments with cylindrical blast wave simulations showed that applying
+    // the same factor to all fields here dramatically improves results for the
+    // cylindrical blast wave. Of course, this might differ depending on the
+    // test problem.
+    constexpr double safety = 0.95;
+    const double factor =
```

Comment:
I don't understand `factor`.  If I have a function X with a mean X_m , and an extremum X_x violating bound X_b then I get the factor should be (X_m - X_b)/(X_m - X_x).  So in this formula you appear to be using a bound of zero, in which case you will be driving TildeD towards zero for positive values of TildeD below the cutoff.  I understand the safety factor.  I don't understand the ratio of the minimum over the cutoff.  I think you should either change line 142 to be min_d is negative or use the cutoff as the X_b and drop the minimum value of D option...  (Note if you use the cutoff as the bound, it also needs to be multiplied by sqrt_det_spatial_metric)